src/porepy/numerics/time_step_control.py

```diff
@@ -0,0 +1,281 @@
+        if recomp_max is not None:
+            self.recomp_max = recomp_max
+        else:
+            self.recomp_max = 10
+
+        # Print information
+        if print_info is None:
+            self._print_info = True
+        else:
+            self._print_info = print_info
+
+        # Time
+        self.time = self.time_init
+
+        # Time step. Initially, equal to the initial time step
+        self.dt = self.dt_init
+
+        # Recompute solution flag
+        self.recomp_sol = False
+
+        # Private attributes
+        # Number of times the solution has been recomputed
+        self._recomp_num = 0
+
+        # Index of the next scheduled time
+        self._scheduled_idx = 1
+
+    def __repr__(self) -> str:
+
+        s = "Time-stepping control object with atributes:\n"
+        s += f"Initial simulation time = {self.time_init}\n"
+        s += f"Final simulation time = {self.time_final}\n"
+        s += f"Initial time step = {self.dt_init}\n"
+        s += f"Minimum time step = {self.dt_min}\n"
+        s += f"Maximum time step = {self.dt_max}\n"
+        s += f"Lower optimal iteration range = {self.iter_low}\n"
+        s += f"Upper optimal iteration range = {self.iter_upp}\n"
+        s += f"Lower multiplication factor = {self.iter_low_factor}\n"
+        s += f"Upper multiplication factor = {self.iter_upp_factor}\n"
+        s += f"Recompute solution multiplication factor = {self.recomp_factor}\n"
+        s += f"Maxximum recomputing attempts = {self.recomp_max}"
+
+        return s
+
+    def next_time_step(self, iters: int) -> Union[float, None]:
+        """
+        Determines the next time step based on the previous amount of iterations needed
+        to reach convergence. If convergence was not achieved, then the time step is
+        reduced by recomp_factor. The time-stepping control routine will recompute the
+        solution recomp_max times. Otherwise, an error will be raised and the simulation
+        stopped.
+
+        Parameters
+        iters (int): Number of non-linear iterations. In time-dependent simulations,
+            this tipically represent the number of iterations for a time step.
+
+        Returns
+        -------
+        dt: float
+            Next time step
+        """
+
+        # First, check if we are allowed to recompute the solution
+        if self._recomp_num > self.recomp_max:
+            s = f"Solution did not convergece after {self.recomp_max}"
+            s += " recomputing attempts."
+            raise ValueError(s)
+
+        # Check if we reach final simulation time
+        if self.time == self.time_final:
+            return None
+
+        # If iters == max_iter:
+        #   Decrease time step by the recomputing factor
+        #   Update time (since solution will be recomputed)
+        #   Set to True the re-computation flag
+        #   Increase counter that keeps track of how many times the solution was recomputed
+        if iters == self.iter_max:
```

Comment:
Does this mean you force recomputation if you happen to converge at the maximum number of permitted iterations? Or is there a 0-offset factor?

Reply:
This is a very good question. As it is, it will force recomputation. I'm going to change this behavior. I will point out the changes in the reviewed version.